`UnpackingZipBasedPackage` und `OPENBLAS_NUM_THREADS`

MHajoha opened the pull request:
Das mit dem `UnpackingZipBasedPackage` hatte ich ursprünglich in https://github.com/questionpy-org/questionpy-server/commit/14e54a47ccbe2cc92123490fa0c90370d48af120 geändert, jetzt auch auf dev.

Und den vorhin erwähnten OpenBLAS/NumPy-Fix. Damit erhält der Worker auch nicht mehr die Umgebungsvariablen des Servers

Fixes: #138

Edit: Ausgehend davon, dem Worker ein eigenes Arbeitsverzeichnis zu geben, in das die Pakete entpackt werden, kamen hier noch einige Änderungen hinzu:
- Um den Worker-Verzeichnissen halbwegs sinnvolle Namen zu geben, kriegen Worker jetzt vom Pool einen fixen Namen.
- Der Name wird jetzt auch als Präfix der Task-Namen (und im Falle des ThreadWorkers des Thread-Namens) verwendet. 
- Alle `BaseWorkerError`-Subklassen geben den Worker-Namen mit an
- Für die bessere Übersicht habe ich die `Worker.__init__`-Argumente in ein `TypedDict` verpackt. Mit Python 3.12 kann `**kwargs` damit typisiert werden, Mypy prüft dann die Zugriffe. Deshalb habe ich die minimale Python-Version auf 3.12 gesetzt.

MHajoha committed:
feat: replace ZipBasedPackage with UnpackingZipBasedPackage

zipimport doesn't support extension modules.

MHajoha committed:
fix: set OPENBLAS_NUM_THREADS to 2 in subprocesses

This keep OpenBLAS from blowing up any package that uses NumPy.

MartinGauk commented on questionpy_server/worker/runtime/package.py:
```diff
@@ -68,38 +71,42 @@ def state(self) -> PackageState:
+    def __del__(self) -> None:
+        if self._dir_package:
+            shutil.rmtree(self._dir_package.path)
+            self._dir_package = None
```
Die Stelle gefällt mir nicht so. Der Worker sollte nicht selber zuständig sein, dass aufgeräumt wird. Das sollte im Server beim Subprocess-Worker geschehen. Bei einem späteren Container-basierten Worker passiert das Aufräumen sicherlich automatisch, wenn wir ein tmpfs mounten.

MHajoha replied:
Vermutlich ergibt es Sinn, jedem Worker ein Verzeichnis zuzuweisen, dass nach Ende des Workers entfernt wird. Ich würde mal folgendes umsetzen:
```
<Pool-Verzeichnis>/
└── worker-0/
    ├── packages/
    │   └── b7e4f42cc6765f8ec679f5ba09c1243d7f8f5f89c39f6b880ff82cd9b8d4e160/
    │       ├── dist/
    │       │   ├── ...
    │       │   └── qpy_manifest.json
    │       ├── ...
    │       └── qpy_config.yml
    └── workdir/ 
```

Für `<Pool-Verzeichnis>` würde ich zunächst `mkdtemp` nehmen. Eine Art `workdir` halte ich für sinnvoll, würde ich aber noch nicht implementieren. Fürs erste würde ich weiterhin entpacken, aver sowas wie [mount-zip](https://github.com/google/mount-zip) halte ich für die Zukunft für interessant.

Was hälst du davon?

MartinGauk replied:
Finde ich gut.

MHajoha committed:
refactor: rename start_server to run_forever

Makes it clear that the method blocks while the server runs.

MHajoha committed:
refactor: use cleanup_ctx to manage collectors and worker pool

on_shutdown hooks run _before_ remaining connections and request handlers are cancelled, and they might require workers and collectors. on_cleanup/cleanup_ctx stops them after connections and request handlers.

MHajoha committed:
chore: bump minimum python version to 3.12

MHajoha committed:
feat: unpack packages in a dir owned by the worker

The entire directory is cleaned up after the worker exits.

MHajoha committed:
feat: use worker name in task and thread names

MHajoha committed:
feat: show worker name in BaseWorkerErrors

MartinGauk commented on questionpy_server/worker/impl/_base.py:
```diff
@@ -187,15 +185,20 @@ async def _observe(self) -> None:
+            if self.worker_home.exists():
+                # Security: rmtree deletes symlinks to dirs without recursing into them, so there should be no danger
+                # of a malicious package causing us to delete anything outside the worker home.
+                shutil.rmtree(self.worker_home)
```
Das Löschen sollte mit `asyncio.to_thread` in einem eigenen Thread geschehen.

MartinGauk commented on questionpy_server/worker/impl/_base.py:
```diff
@@ -339,6 +342,12 @@ async def get_static_file_index(self) -> dict[str, PackageFile]:
+    def __del__(self) -> None:
+        if self.state != WorkerState.NOT_RUNNING:
+            log.warning("Worker '%s' was not stopped correctly.", self.name)
+        elif self.worker_home.exists():
```
Das müsste eigentlich auch in einem anderen Thread laufen. Ich denke, auf die Überprüfung kann aber auch verzichtet werden.

MHajoha replied:
Ja stimmt. Async-Sachen in `__del__` sind vermutlich nicht safe...
Dann lasse ich es weg

MartinGauk commented on questionpy_server/worker/pool.py:
```diff
@@ -157,6 +193,21 @@ async def _free_memory(self, required_memory: int) -> None:
+        else:
+            assert_never(package)
+
+        random_part = b32hexencode(self._random.randbytes(5)).lower().decode("ascii")
```
Warum nicht einfach durchnummerieren?

MHajoha replied:
Das würde Paketen Aufschluss darüber erlauben, wie viele Worker gerade laufen oder seit Server-Start liefen. Das ist vermutlich nicht wirklich schlimm, fand ich aber unschön. Wir können auch gerne einen counter verwenden, da habe ich keine starke Meinung zu.

MartinGauk replied:
Ich bin für den Counter als einfache Lösung, um dem (wenn auch geringen) Risiko von Kollisionen aus den Weg zu gehen.

janbritz replied:
Als Kompromiss könnten wir auch auf `uuid.uuid4()` zurückgreifen.

MartinGauk commented on questionpy_server/worker/pool.py:
```diff
@@ -174,7 +225,11 @@ async def _create_or_reuse_worker(
+
+            name = self._generate_worker_name(package, lms, context)
+            worker_home = self._working_dir / f"worker-{name}"
+            worker_home.mkdir()
```
Das müsste auch in einem anderen Thread geschehen.

MartinGauk commented on questionpy_server/worker/runtime/package.py:
```diff
@@ -68,38 +69,34 @@ def state(self) -> PackageState:
+    def _unpack(self, to_dir: Path) -> "DirBasedPackage":
+        to_dir.mkdir(parents=True)
+        with ZipFile(self.path) as zip_file:
+            zip_file.extractall(to_dir)
```
Am besten sollte nur der `DIST_DIR` Ordner entpackt werden.

MartinGauk commented on questionpy_server/worker/pool.py:
```diff
@@ -47,6 +62,14 @@ def __init__(self, max_workers: int, max_memory: int, worker_type: type[Worker]
         self.max_memory = max_memory
 
+        # TODO: Make this configurable (#137)
+        self._limits_per_worker = WorkerResourceLimits(max_memory=200 * MiB, max_cpu_time_seconds_per_call=10)
```
Das soll schon in get_worker bleiben. Die Limits sollen nicht fest sein, siehe https://github.com/questionpy-org/questionpy-server/issues/137#issuecomment-2809907297

MHajoha replied:
Ich habe diese Limits in eine Konstante gezogen, den Check zurück in `get_worker` (wobei ich denke, dass der dort mittelmäßig viel Sinn macht) und den Kommentar ausführlicher gemacht.

MartinGauk commented on questionpy_server/worker/__init__.py:
```diff
@@ -50,13 +51,27 @@ class PackageFileData:
+    package: PackageLocation
+    """The main package that the worker should load when [start][questionpy_server.worker.Worker.start] is called."""
+    worker_home: Path
+    """An existing directory owned by the worker, with the same lifetime as the worker."""
```
Im Falle eines Container-basierten Workers brauchen wir den Ordner nicht. Wie sollten wir damit umgehen?

MHajoha replied:
Ich denke im Detail sollte das warten, bis wir tatsächlich Container-basierte Worker implementieren. Vielleicht ist das `worker_home` da dann einfach `/questionpy` oder so, vgl. mit GitHub's `/github`: https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#docker-container-filesystem

MartinGauk replied:
Ja, wir warten bis dahin.

Mir fällt auch ein, dass wir diesen Ordner nutzen können, um darin Named Pipes zu erstellen, die mit dem stdin/out/err anderer Prozesse verbunden sind. Es ist ja angedacht, dass ein Paket einen weiteren Container starten lassen können soll, in dem z.B. eine Code-Abgabe kompiliert und getestet/ausführt wird.

Den Container-Workern würde ich aber keinen Schreibzugriff auf diesen Ordner geben wollen. Die sollen nur in ein jeweils gemountetes tmpfs schreiben können.

MHajoha committed:
feat: extract only dist dir

MHajoha committed:
fix: no IO in main thread

MHajoha committed:
refactor: _DEFAULT_LIMITS constant; explain why the check is there

MHajoha committed:
fix: graceful handling of pipe EOF on write

MHajoha committed:
fix: work around unclosed aiohttp transports

This caused ResourceWarnings in tests.

See: https://docs.aiohttp.org/en/stable/client_advanced.html#graceful-shutdown

janbritz commented on questionpy_server/worker/impl/_base.py:
```diff
@@ -187,21 +189,24 @@ async def _observe(self) -> None:
+
     async def stop(self, timeout: float) -> None:
-        try:
+        # If these are raise, either the worker is already dead, in which case we're happy, or it's somehow broken,
```
```suggestion
        # If these exceptions are raised, either the worker is already dead, in which case we're happy, or it's somehow broken,
```

MartinGauk commented on questionpy_server/worker/runtime/package.py:
```diff
@@ -68,38 +69,36 @@ def state(self) -> PackageState:
+        to_dir.mkdir(parents=True)
+        with ZipFile(self.path) as zip_file:
+            for info in zip_file.infolist():
+                if info.filename.startswith(f"{DIST_DIR}/"):
```
Ich befürchte, dass der String immer wieder neu gebaut wird. Da das ZIP durchaus einige Dateien beinhalten könnte, lohnt es sich etwas, den String vor der Schleife zu bilden.

MHajoha committed:
perf: build dist prefix only once

MHajoha committed:
refactor: enumerate workers
